thrift-logger/src/main/java/com/pinterest/singer/client/AuditLogbackThriftLogger.java

```diff
@@ -0,0 +1,161 @@
+
+  private static final String AUDIT_THRIFT_LOGGER_COUNT_METRIC = "audit.thrift_logger.count";
+  private static final String AUDIT_THRIFT_LOGGER_ERROR_TEXCEPTION = "audit.thrift_logger.error.texception";
+  private static final String AUDIT_THRIFT_LOGGER_ERROR_LOGBACKEXCEPTION = "audit.thrift_logger.error.logbackexception";
+  private static final String AUDIT_THRIFT_LOGGER_ERROR_INIT = "audit.thrift_logger.error.init";
+  private static String HOST_NAME;
+
+  static {
+    try {
+      String hostName = InetAddress.getLocalHost().getHostName();
+      int firstDotPos = hostName.indexOf('.');
+      if (firstDotPos > 0) {
+        HOST_NAME = hostName.substring(0, firstDotPos);
+      }
+    } catch (Exception e) {
+      // fall back to env var.
+      HOST_NAME = System.getenv("HOSTNAME");
+    }
+  }
+
+  private final Appender<LogMessage> appender;
+  private final String topic;
+  private Class<?> thriftClazz;
+  private TFieldIdEnum loggingAuditHeadersField;
+  private AuditHeadersGenerator auditHeadersGenerator;
+
+  public void setAuditHeadersGenerator(AuditHeadersGenerator auditHeadersGenerator) {
+    this.auditHeadersGenerator = auditHeadersGenerator;
+  }
+
+  public AuditLogbackThriftLogger(Appender<LogMessage> appender, String topic,
+                                  Class<?> thriftClazz) {
+    if (thriftClazz == null) {
+      throw new IllegalArgumentException(
+          "To initialize AuditLogbackThriftLogger, thrift class must be valid class object.");
+    }
+    this.appender = appender;
+    this.topic = topic;
+    this.auditHeadersGenerator = new AuditHeadersGenerator(HOST_NAME, topic);
+    this.thriftClazz = thriftClazz;
+    init(this.thriftClazz);
+  }
+
+  public void init(Class<?> thriftClazz) {
+    try {
+      Map m = (Map) thriftClazz.getDeclaredField("metaDataMap").get(null);
+      for (Object o : m.keySet()) {
+        TFieldIdEnum tf = (TFieldIdEnum) o;
+        if ("loggingAuditHeaders".equalsIgnoreCase(tf.getFieldName())) {
+          this.loggingAuditHeadersField = tf;
+          LOG.info("Found loggingAuditHeaders field for {}", thriftClazz.getName());
+          break;
+        }
+      }
+      if (this.loggingAuditHeadersField == null) {
+        throw new NoSuchFieldException(
+            "Cannot find loggingAuditHeaders field for " + thriftClazz.getName());
+      }
+    } catch (Exception e) {
+      OpenTsdbMetricConverter.incr(AUDIT_THRIFT_LOGGER_ERROR_INIT, "topic=" + topic,
+          "host=" + HOST_NAME);
+      LOG.error("Init method cannot finish successfully due to {}, needs to exit.", e.getMessage());
+      System.exit(1);
+    }
+  }
+
+  @Override
+  public void append(byte[] message) {
+    throw new UnsupportedOperationException("For AuditLogbackThriftLogger, "
+        + "please use append(TBase thriftMessage))");
+  }
+
+  @Override
+  public void append(byte[] partitionKey, byte[] message, long timeNanos) {
```

Comment:
Why are we not supporting this?

Reply:
We would like to inject LoggingAuditHeaders to the original thrift message, as its optional field.  If thrift message is passed as the argument, it is much efficient as we can set the LoggingAuditHeaders field directly; in the contrast, if message is passed as byte[], then we need to deserialize first.   If any applications use AuditLogbackThriftLogger, we want to enforce that the application is using **append(byte[] partitionKey, TBase message, long timeNanos)** rather than this one.

Reply:
The message bytes are wrapped in our custom thrift object so why does the payload matter?

Reply:
This method will be called by various applications that use thrift logger to log message to files.